lib/image-util.js

```
@@ -0,0 +1,118 @@
+    .on('data', (chunk) => chunks.push(chunk)).on('end', () => {
+      resolve(Buffer.concat(chunks).toString('base64'));
+    })
+    .on('error', (err) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      reject(err);
+    });
+  });
+}
+
+/**
+ * Crop the image by given rectangle
+ *
+ * @param {PNG} image The image to mutate by cropping
+ * @param {Region} rect The selected region of image
+ */
+function cropImage (image, rect) {
+  const imageRect = {width: image.width, height: image.height};
+  const interRect = getRectIntersection(rect, imageRect);
+  if (isEmptyRect(interRect)) {
+    throw new Error(`Cannot crop ${JSON.stringify(rect)} from ${JSON.stringify(imageRect)} because the intersection between them is an empty rectangle`);
+  }
+
+  const firstVerticalPixel = interRect.top;
+  const lastVerticalPixel = interRect.top + interRect.height;
+
+  const firstHorizontalPixel = interRect.left;
+  const lastHorizontalPixel = interRect.left + interRect.width;
+
+  const croppedArray = [];
+  for (let y = firstVerticalPixel; y < lastVerticalPixel; y++) {
+    for (let x = firstHorizontalPixel; x < lastHorizontalPixel; x++) {
+      const firstByteIdxInPixelBlock = (imageRect.width * y + x) << 2;
+      for (let byteIdx = 0; byteIdx < BYTES_IN_PIXEL_BLOCK; byteIdx++) {
+        croppedArray.push(image.data[firstByteIdxInPixelBlock + byteIdx]);
+      }
+    }
+  }
+
+  image.data = new Buffer(croppedArray);
+  image.width = interRect.width;
+  image.height = interRect.height;
+  return image;
+}
+
+function getRectIntersection (rect, imageSize) {
+  const left = rect.left >= imageSize.width ? imageSize.width : rect.left;
+  const top = rect.top >= imageSize.height ? imageSize.height : rect.top;
+
+  // if we can't find a total intersection of rect with imageSize, then say
+  // that width and/or height are 0, to indicate we didn't get a valid
+  // intersection
+  const width = imageSize.width >= (left + rect.width) ? rect.width : 0;
```

I'm afraid this may not work properly if the intersection is not full `(left < imageSize.width && left + rect.width > imageSize.width)`. The resulting `width` should not be zero in  such case

---

I think the point is that we do not allow non-full intersections, and setting the width to 0 is a way of doing that. We could also throw an error instead?

---

Yes @jlipps, you are right, we don't need to support partial intersections (for current use-case).
Also, we already have an error which will be thrown in such case (on line 77), but to be more clear, we can add addition one inside getRectIntersection.

---

Then we need to adjust the function name, otherwise it's misleading.

---

I'd expect partial intersections to be calculated properly, since it's a pretty valid case

---

ok @mykola-mokhnach @astappev i changed things around so the intersection logic itself is more general and the calling function does the check and the exception to keep the current overall behavior. what do you think?